### PR TITLE
feat(perf): add workload_trace dataset plugin for production traffic replay

### LINKED ISSUE #1489: [Feature] perf: add workload_trace dataset plugin for production traffic replay
## Feature Description

Add a `workload_trace` dataset plugin that replays recorded production traffic traces using their original arrival pattern (open-loop), enabling realistic inference performance benchmarking.

The plugin targets **stateless, open-loop replay** of recorded requests — each trace record is sent as an independent request. The request body may contain **multi-turn** conversation history (e.g., a messages array with multiple roles), but the plugin does not maintain cross-request session state.

## Background

Existing datasets (`random`, `line_by_line`, etc.) generate synthetic workloads that do not fully capture production traffic characteristics: bursty arrival patterns, varying prompt lengths, and custom headers for A/B testing / canary routing.

The `trie_*` plugins handle multi-turn agent traces; `workload_trace` complements them by supporting open-loop replay of arbitrary recorded traffic.

## Expected Behavior

- **Input**: JSONL file, each line a record with `body` (required), `timestamp` (required), `headers` (optional), `request_id` (optional).
- **Replay**: Requests are dispatched based on relative timestamp offsets, with a `speed` multiplier controlling replay rate (1.0 = original, 2.0 = 2× faster, 0.5 = 2× slower). Timestamps must be monotonically non-decreasing; out-of-order records should fail fast with a clear error.
- **Model rewriting**: Optional `model_override` (global) and `model_mapping` (per-model) args via either a new `--dataset-args` or the existing `--multi-turn-args` (requires a relaxed schema). A mapping match takes priority; `model_override` is the fallback when no mapping entry matches.
  > Note: the CLI `--model` argument is required but does **not** override a trace body's existing `model` field (due to `setdefault` semantics from #1485). Use `model_override` or `model_mapping` (via `--dataset-args` or `--multi-turn-args`) for explicit model rewriting.
- **Per-request headers**: Merged with the benchmark client's global headers. Per-request headers take precedence on key conflict. Hop-by-hop headers (`Content-Length`, `Host`, `Connection`) should be stripped.
- **Request ID propagation**: If `request_id` is present in the trace, it should be propagated to benchmark results so users can correlate benchmark output with the original trace.
- **Validation**: Malformed JSON, missing required fields, type errors, reserved key collisions, and out-of-order timestamps should all fail fast with a clear error message and line number.
- **Open-loop only**: The plugin owns the request arrival schedule and requires `--open-loop`. Incompatible configurations (closed-loop, rate-generation modes) should fail fast rather than silently changing replay semantics.

A possible usage example:

```bash
evalscope perf \
  --dataset workload_trace \
  --dataset-path trace.jsonl \
  --multi-turn-args '{"speed": 2.0, "model_mapping": {"qwen-72b": "qwen-72b-v2"}}' \
  --model qwen-72b-v2 \
  --url http://target:8080/v1/chat/completions \
  --open-loop
```

> By default the **entire trace** is replayed; `--number` and `--rate` are not needed (the arrival schedule is derived from trace timestamps).

### PR DESCRIPTION
**Pull request by qiumuyang**
## Summary

Add a `workload_trace` dataset plugin that replays recorded production traffic with original timing, headers, and request bodies — enabling performance benchmarking against real-world workload patterns.

Closes #1489

## Motivation

Existing perf datasets (`openqa`, `share_gpt`, `random`, etc.) generate synthetic traffic that may not reflect production load characteristics. Real production workloads have bursty arrival patterns, heterogeneous request shapes, and model-specific routing. This plugin bridges that gap by replaying captured JSONL traces verbatim.

### New files

| File | Description |
|------|-------------|
| `evalscope/perf/plugin/datasets/workload_trace.py` | Core plugin (249 lines) |
| `evalscope/perf/utils/body_meta.py` | Sideband key constants for per-request metadata |
| `tests/perf/test_workload_trace.py` | 48 unit tests |
| `tests/perf/test_workload_trace_e2e.py` | 7 E2E tests (mock HTTP server) |

### Modified files

| File | What changed |
|------|-------------|
| `arguments.py` | `--dataset-args` CLI arg; `--model` / `--number` optional for trace datasets; `requires_model` validation |
| `benchmark.py` | Re-read `total_count` / `warmup_count` after plugin construction (plugins may adjust them); remove unused `import numpy` |
| `plugin/api/base.py` | `extract_body_meta()` — pops sideband keys before `process_request`; merges per-request headers with fill semantics (CLI headers win) |
| `plugin/api/openai_api.py` | Guard `model=None` when dataset carries model in body |
| `plugin/datasets/base.py` | `provides_arrival_schedule` / `requires_model` class attributes |
| `plugin/datasets/__init__.py` | Register import |
| `core/strategies/open_loop.py` | Trace-offset dispatch branch with deferred duration deadline for warmup |
| `core/http_client.py` | Call `extract_body_meta` before delegating to API plugin |

## Trace format

Each line in the JSONL file:

```jsonl
{"body": {"model": "qwen-plus", "messages": [...]}, "timestamp": 1700000000.0}
{"body": {"model": "qwen-max", "messages": [...]}, "timestamp": 1700000001.5, "headers": {"X-Tag": "exp"}, "request_id": "req-42", "completion_tokens": 256}
```

- **`body`** (dict or JSON string): complete request body, sent as-is.
- **`timestamp`** (number or ISO-8601): arrival time; only relative deltas matter.
- **`headers`** (optional): per-request HTTP headers (merged with fill semantics — CLI headers take precedence).
- **`request_id`** (optional): propagated to results for correlation.
- **`completion_tokens`** (optional): used by `match_output_length`.

## Usage

```bash
# Basic replay
evalscope perf \
  --dataset workload_trace \
  --dataset-path trace.jsonl \
  --url http://localhost:8000/v1/chat/completions \
  --open-loop

# 2× speed, remap models, match recorded output lengths
evalscope perf \
  --dataset workload_trace \
  --dataset-path trace.jsonl \
  --url http://localhost:8000/v1/chat/completions \
  --open-loop \
  --dataset-args '{"speed": 2.0, "model_mapping": {"gpt-4": "qwen-max"}, "match_output_length": true}'

# Limit to first 500 requests
evalscope perf \
  --dataset workload_trace \
  --dataset-path trace.jsonl \
  --url http://localhost:8000/v1/chat/completions \
  --open-loop \
  --number 500
```

### Dataset args (`--dataset-args`)

| Key | Type | Default | Description |
|-----|------|---------|-------------|
| `speed` | float | `1.0` | Replay speed multiplier (2.0 = 2× faster) |
| `model_override` | str | `None` | Replace all models with this value |
| `model_mapping` | dict | `None` | Selectively remap model names |
| `match_output_length` | bool | `false` | Set `max_tokens` from `completion_tokens` and enable `ignore_eos` (requires vLLM or compatible) |

## Design decisions

- **Sideband keys**: per-request metadata (`__evalscope_*` prefixed) is embedded in the body dict and stripped by `extract_body_meta` before reaching API plugins. This avoids changing the plugin interface.
- **Fill semantics for headers**: `{**trace_headers, **cli_headers}` — CLI-set headers (e.g. `Authorization` from `--api-key`) always take precedence over trace headers, so explicit CLI configuration is never silently overridden by trace data.
- **`--model` / `--number` optional**: Trace data carries its own model and count. New `requires_model` class attribute on `DatasetPluginBase` controls validation.
- **Deferred deadline**: `--duration` countdown starts after warmup completes, so warmup doesn't consume benchmark time.

## Testing

```
$ pytest tests/perf/test_workload_trace.py tests/perf/test_workload_trace_e2e.py -q
55 passed
```

Coverage includes: loading/validation, timestamp parsing (epoch + ISO-8601), body-as-string, model rewriting, speed scaling, warmup, `--number` truncation, `match_output_length`, header fill semantics, reserved key rejection, E2E body passthrough, and E2E warmup exclusion.

### CODE REVIEW

File: evalscope/perf/plugin/datasets/workload_trace.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    def _parse_body(cls, v: Any) -> Dict[str, Any]:
+        if isinstance(v, str):
+            try:
+                parsed = json.loads(v)
+            except json.JSONDecodeError as e:
+                raise ValueError(f'body is a string but not valid JSON: {e}')
+            if not isinstance(parsed, dict):
+                raise ValueError(f'body JSON string must decode to a dict, got {type(parsed).__name__}')
+            return parsed
+        return v
```
[the remark this answers is not in the document]
Comment (qiumuyang) [reply evalscope/perf/plugin/datasets/workload_trace.py:110]: Empirically the pydantic-to-dict switch saves little end-to-end for a one-time startup cost, and the refactor would lose Pydantic's built-in validation. Tracked as a potential follow-up if traces reach millions of records.

File: evalscope/perf/core/strategies/open_loop.py [hunk 2]
```diff
@@ -89,7 +90,40 @@ async def _run_phase(self, requests: List[dict], is_warmup: bool, deadline: Opti
         n = len(requests)
         rate = self.args.rate
 
-        if rate == -1 or n == 0:
+        # Check for pre-computed arrival offsets.
+        has_trace_offsets = n > 0 and BODY_META_ARRIVAL_OFFSET in requests[0]
+
+        if has_trace_offsets:
+            delay_ts = np.array([r.pop(BODY_META_ARRIVAL_OFFSET) for r in requests])
```
[the remark this answers is not in the document]
Comment (qiumuyang) [reply evalscope/perf/core/strategies/open_loop.py:99]: This key is unconditionally set by the sole producer (workload_trace.build_messages), so the pop never raises on the trace-offsets path; a silent 0.0 default would mask a real corruption rather than surface it.

File: evalscope/perf/plugin/datasets/workload_trace.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        if not self._records:
+            raise ValueError('workload_trace: dataset is empty')
+
+        for i in range(1, len(self._records)):
+            if self._records[i].timestamp < self._records[i - 1].timestamp:
+                raise ValueError(
+                    f'Timestamps must be monotonically non-decreasing: '
+                    f'record {i + 1} (t={self._records[i].timestamp}) < '
+                    f'record {i} (t={self._records[i - 1].timestamp})'
+                )
```
[the remark this answers is not in the document]
Comment (qiumuyang) [reply evalscope/perf/plugin/datasets/workload_trace.py]: I've deduplicated the manual pre-checks that overlapped with Pydantic validators.

File: tests/perf/test_workload_trace.py [hunk 4]
```diff
@@ -0,0 +1,514 @@
+"""Tests for workload_trace dataset plugin."""
+
+import json
+import numpy as np
+import os
+import pytest
```
[the remark this answers is not in the document]
Comment (qiumuyang) [reply tests/perf/test_workload_trace.py:6]: Passes pre-commit run isort with the project's pinned config (v7.0.0 + setup.cfg).

File: evalscope/perf/plugin/api/base.py [hunk 5]
```diff
@@ -12,6 +13,28 @@ def __init__(self, param: Arguments) -> None:
...
+        Returns:
+            (headers, request_id)
+        """
+        extra_headers = body.pop(BODY_META_HEADERS, None)
+        request_id = body.pop(BODY_META_REQUEST_ID, None)
+        if extra_headers:
+            headers = {**extra_headers, **headers}
+        for k in [k for k in body if k.startswith(BODY_META_PREFIX)]:
+            body.pop(k)
+        return headers, request_id
```
[the remark this answers is not in the document]
Comment (qiumuyang) [reply evalscope/perf/plugin/api/base.py:52]: Hop-by-hop stripping added in 4d9d9ee. Fill semantics (CLI wins) is intentional — prevents trace data from overriding CLI-set `Authorization` from `--api-key`.

File: evalscope/perf/plugin/datasets/workload_trace.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+            n = min(query_parameters._current_number(), n)
+            self._records = self._records[:n]
+            self._schedule = self._schedule[:n]
+        query_parameters.number = n
+
+        # --model: if user explicitly passed it, treat as model_override fallback.
+        if 'model' in query_parameters.model_fields_set:
+            if not self.dataset_args.model_override and not self.dataset_args.model_mapping:
+                self.dataset_args.model_override = query_parameters.model
+
```
[the remark this answers is not in the document]
Comment (qiumuyang) [reply evalscope/perf/plugin/datasets/workload_trace.py]: This is a convenience fallback — only activates when neither `model_override` nor `model_mapping` is set. Now logs explicitly when it takes effect (4d9d9ee).

File: evalscope/perf/core/http_client.py [hunk 7]
```diff
@@ -78,8 +78,11 @@ async def post(self, body) -> BenchmarkData:
             BenchmarkData: The benchmark data object containing request and response information.
         """
         try:
+            headers, request_id = self.api_plugin.extract_body_meta(body, self.headers)
             # Delegate the request processing to the API plugin
-            output = await self.api_plugin.process_request(self.client, self.url, self.headers, body)
+            output = await self.api_plugin.process_request(self.client, self.url, headers, body)
+            if request_id:
+                output.trace_id = request_id
```
[the remark this answers is not in the document]
Comment (qiumuyang) [reply evalscope/perf/core/http_client.py]: Fixed in 4d9d9ee — added a dedicated `request_id` field on `BenchmarkData` and DB schema instead of reusing `trace_id`.

File: evalscope/perf/core/strategies/open_loop.py [hunk 8]
```diff
@@ -89,7 +90,40 @@ async def _run_phase(self, requests: List[dict], is_warmup: bool, deadline: Opti
...
+                        )
+                        break
+                    sleep_s = target_times[i] - time.perf_counter()
+                    if actual_deadline is not None:
+                        sleep_s = min(sleep_s, actual_deadline - time.perf_counter())
+                if sleep_s > 0:
+                    await asyncio.sleep(sleep_s)
+                task = asyncio.create_task(
+                    _send_request_open_loop(request, is_warmup or is_warmup_req, self.queue, self.client)
+                )
```
[the remark this answers is not in the document]
Comment (qiumuyang) [reply evalscope/perf/core/strategies/open_loop.py:128]: I think it's fine for `--duration` as a soft deadline, but happy to harden if needed.

**Review by Yunnglin — APPROVED** [review]
Pushed a few review-driven refinements on top of the latest commit:

1. **`--model` no longer rewrites trace bodies** (per #1489). The previous fallback silently turned `--model X` into `model_override`, collapsing multi-model traces into one. It now just logs a warning and each request keeps its own `model`; use `--dataset-args` `model_override`/`model_mapping` for explicit rewriting.
2. **Guarded `model=None` in the remaining API plugins** (dashscope / responses / embedding / rerank / custom). Only `openai_api` had the guard, so the others would write `model: null` when `--model` is omitted.
3. **Gated per-request `update_gpu_usage()` to `--api local`.** For remote APIs and `local_vllm` the client process holds no CUDA memory (measurement is always 0), and the first-request `import torch` (~3.5–6s) blocked the event loop — which distorted open-loop **trace arrival timing** for compact traces. Verified: a 3-request trace with 0.5s/0.8s gaps now replays within ~40ms instead of bursting.
4. **Docs + test**: added the `workload_trace` dataset mode, `dataset-args`, and usage examples (zh/en); added a test asserting `--model` does not rewrite the body.

E2E-verified against a mock OpenAI server: multi-model routing preserved, per-request headers applied, `request_id` propagated to the DB, and no `__evalscope_*` sideband keys leak into outgoing requests.
